lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,182 @@
+require 'puppet/state_machine/state'
+
+module Puppet
+  class StateMachine
+    # The MachineBuilder class provides a DSL for implementing the states and transitions of a state machine,
+    # and validating that the machine is well behaved.
+    class MachineBuilder
+      # @param machine_name [String] The name of the machine to build.
+      def initialize(machine_name)
+        @machine_name = machine_name
+        @states = {}
+        @start_state = nil
+      end
+
+      # Build and verify a new state machine.
+      # @return [Puppet::StateMachine]
+      def build(&blk)
+        yield self
+        verify!
+        Puppet::StateMachine.new(@machine_name, @states, @start_state)
+      end
+
+      # Compose multiple state machines into a a single state machine.
+      #
+      # @example
+      #   machine.compose({machine1: m1, machine2: m2}) do |m|
+      #     m.start_state([:m1, :start])
+      #
+      #     m.transition(
+      #       source: [:m1, :final],
+      #       target: [:m2, :start],
+      #       on: Puppet::StateMachine::FINAL_EVENT)
+      #
+      #     m.final([:m2, :final])
+      #   end
+      def compose(machines, &blk)
+        machines.each_pair do |ns, machine|
+          @states.merge!(machine.namespaced_states(ns))
+        end
+        build(&blk)
+      end
+
+      # Verify that the machine is well defined.
+      # @return [void]
+      def verify!
+        verify_start_state!
+        verify_transition_targets!
+        verify_exiting_transitions!
+      end
+
+      # Define the start state for this state machine.
+      # @param name [Symbol]
+      # @return [void]
+      def start_state(name)
+        @start_state = name
+      end
+
+      # Generate a new state.
+      #
+      # @example
+      #   machine.state(:create_lockfile,
+      #     action: -> { Puppet::Util::Lockfile.new("/var/lock/puppet.lock").lock },
+      #     event: ->(has_lock) { has_lock ? :got_lock : :already_locked },
+      #     transitions: {
+      #       got_lock: :run_agent,
+      #       already_locked: :cancel_run,
+      #     })
+      #
+      # @return [Puppet::StateMachine::State] The generated machine state
+      def state(state_name, opts = {})
+        unhandled = opts.keys - [:action, :event, :type, :transitions]
```

Comment:
It seems odd to me that the various validations in this method (particularly knowledge of the supported state options) exist in the MachineBuilder class - why aren't they part of the State class?